### PR TITLE
Fix for anonymous typed functions

### PR DESCRIPTION
**Pull request by Dveim**
Now tests like q"(..$params) => $expr" are passing successfully.

### CODE REVIEW

**Comment by xeno-by** [comment]
The test failure is spurious, so we shouldn't worry much about it - I'll test everything manually before merging.

**Comment by xeno-by** [comment]
@Dveim Could you sign https://www.typesafe.com/contribute/cla/scala, so that we could get going? In order to facilitate our collaboration, I didn't mention that in the first PR that you submitted, but this time let's do it right.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 1]
```diff
@@ -1355,12 +1362,19 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
-      // may be impossible to distinguish from a self-type and so remains an error.  (See #1564)
+
+      t = t match {
+        case q1 @ Term.Quasi(r1, q2 @ Term.Quasi(r2, tree)) =>
+          atPos(q1, q1)(Term.Tuple(List(atPos(q2, q2)(Term.Quasi(q1.rank, q1.tree)))))
+        case _ => t
+      }
+
+      lazy val isInBraces = t.tokens.nonEmpty && t.tokens.head.is[`(`] && t.tokens.last.is[`)`]
       def lhsIsTypedParamList() = t match {
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Weren't you planning to rename this method?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 2]
```diff
@@ -218,14 +226,27 @@ class QuasiquoteSuite extends FunSuite {
...
 //  }
 
-//  test("""q"(i: Int) => 42" """) {
-//    q"(i: Int) => 42"
-//  }
-//
-//  test("q\"(..$params) => $expr\"") {
-//    val q"(..$params) => $expr" = q"(x: Int, y: String) => 42"
+  test("""q"(i: Int) => 42" """) {
+    q"(i: Int) => 42" // should compile
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: How about checking `show[Syntax]` here?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 3]
```diff
@@ -218,14 +226,27 @@ class QuasiquoteSuite extends FunSuite {
...
+  }
+
+
+  test("q\"(..$params) => $expr\"") {
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Double newline is excessive here.

**Comment by Dveim** [comment]
Done.

2015-07-04 10:28 GMT+03:00 Eugene Burmako <notifications@github.com>:

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 4]
```diff
@@ -1355,12 +1362,19 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
-      // in order to allow anonymous functions as statements (as opposed to expressions) inside
-      // templates, we have to disambiguate them from self type declarations - bug #1565
-      // The case still missed is unparenthesized single argument, like "x: Int => x + 1", which
-      // may be impossible to distinguish from a self-type and so remains an error.  (See #1564)
+
+      t = t match {
+        case q1 @ Term.Quasi(r1, q2 @ Term.Quasi(r2, tree)) =>
+          atPos(q1, q1)(Term.Tuple(List(atPos(q2, q2)(Term.Quasi(q1.rank, q1.tree)))))
+        case _ => t
+      }
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Shouldn't this transformation be applied only is `isInBraces` is true?

**Comment by xeno-by** [comment]
@Dveim Congrats with the first complicated pull request! I've added some comments to your code. After you address them, could you rebase the PR so that it has only one commit per piece of functionality + so that all commits have sharp titles and descriptions that briefly motivate them?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 1 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: At the moment name is correct. As far as I understood, fix for tuples with not only Term.Ascribe (like `(x: X, y, z: Z)`) is not intended to be included in this PR.

Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Okay, we can have that fix in a separate PR.

**Comment by Dveim** [comment]
Sorry, I broke whole structure with git rebase. I'll repair it and make new PR soon.

**Pull request closed without merging**